specification/storagecache/resource-manager/Microsoft.StorageCache/stable/2024-03-01/amlfilesystem.json

```diff
@@ -0,0 +1,1453 @@
+        "operationId": "amlFilesystems_Archive"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.StorageCache/amlFilesystems/{amlFilesystemName}/cancelArchive": {
+      "post": {
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/AmlFilesystemNameParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "AML file system has started the cancellation of the archive operation. Poll the AML file system's hsm.archiveStatus field to monitor the progress."
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "storagecache.json#/definitions/CloudError"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "amlFilesystems_cancelArchive": {
+            "$ref": "./examples/amlFilesystems_CancelArchive.json"
+          }
+        },
+        "tags": [
+          "amlFilesystems"
+        ],
+        "description": "Cancel archiving data from the AML file system.",
+        "operationId": "amlFilesystems_CancelArchive"
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.StorageCache/amlFilesystems/{amlFilesystemName}/import": {
+      "post": {
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v5/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v5/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v5/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/AmlFilesystemNameParameter"
+          },
+          {
+            "name": "ImportInfo",
+            "in": "body",
+            "description": "Information about the import operation",
+            "schema": {
+              "$ref": "#/definitions/AmlFilesystemImportInfo"
+            }
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "AML file system has started the import operation. Poll the AML file system's hsm.importStatus field to monitor the progress."
+          },
```

Comment:
why not make this operation long running instead of asking the user to look at some other object for status ?

Reply:
The api imports information from large storage accounts that could take days.   The job will complete and go away when it's done.  

If this is a resource, then the customer has to delete and cleanup old ones.  There are no settings to change other than to cancel it to start a new one.   The status is an array in case in the future we would have the ability to have parallel import jobs, so the provided key was added as the x-ms-identifier instead of trying to use the array of import prefixes as a potential key.  Right now there is only one, but the idea was that it would be a breaking change to make it an array, so we went with an array to start.    

What is the limit how long an async op can run?  This could take a while depending on how large the dataset they want to import is?

Should we use a sub resource for jobs that would just go away?  If it's a sub resource we can't auto delete them, can we?
At some point I think we would want to block additional imports if they haven't cleaned out the old ones.

Reply:
My understanding from above is:

When you define it as an action named "Import", and it returns 200 response, it could be interpreted as that "Import has completed".

How about you model it an Post Action named "startImportJob"? Then, as soon as your backend is able to start the job successfully, you return 200 response to user. The job can continue to work and complete the import behind the scenes. For user's tracking, the 200 response of "startImportJob" gives the user an "operationId", using which the user can track the status of import operation (e.g. 10% complete or fully completed etc.).

Even better will be model it as PUT "/importJobs/{jobName}", and user can provide some name to job and does a PUT to the job. When the job is created successfully, PUT returns with 201 response, along with an operationId for user to track the status of import.

The importJobs can be a proxy resource, with only GET, PUT defined. If delete is not required, don't implement DELETE API.

What do you think of these suggestions?

Tagging @raosuhas for his view points also.

Reply:
@visingla-ms,

Here's some additional information.

There is an import status being added in the amlfilesystem resource that has percent complete, detailed errors, start and completion time, and state. It's all readonly and can be shown in the GET.  

The previous versions did an import as part of PUT supplying the container to import from and any optional filesystem paths to get within the container.  That is still available and kicks off the import as part of the initial PUT to import data into the cluster.   This change allows the customer to import additional information after the initial PUT which wasn't possible before.  For backward compatibility we still need to import during the initial creation.

In the current version, the amlfilesystem PUT waits on the import to complete during provisioning of the cluster.  This can make just building the cluster take a long time because it included the import.  Now the import would only kicked off from PUT and the status is provided to watch for its completion and for status of additional imports.

The importStatus is an array instead of a single object to handle the future case of potentially handling parallel imports from different filesystem paths.  We currently don't support that so there would only be 1 at a time.   The key is the array of filesystem paths to pull from, which is complex to use as a key.  The unique id was used to key the potential parallel imports for the importStatus array if that becomes available.

1.  Add LRO details to the post operations and update the name.
If we use Post Action startImportJob, its similar to what we have.
The long running operation id can be returned with status.
Would you expect the detailed importStatus to go away?  It covers the status of the import kicked off during the initial PUT amlfilesystem.  It also has all of the details and can always be found by using list or get on your amlfilesystems in case you lose the async op id given that these may take hours to days.

2. If we use the proxy resource with only GET/PUT,
The resource isn't tied in the swagger spec to amlfilesystem directly.  Its floating in with other things like operations and async operations apis making it more difficult to find in the docs and link to the amlfilesystem object that the job is for.
The customer has to monitor that resource for status using its GET.   If this is a separate resource then they can't just see how the import is going on this cluster with GET on amlfilesystem.  Instead they would have to monitor GET on another resource.   
Would the importStatus we have in the amlfilesystem go away or also still exist?
Would we need a collection GET to see them all since its a separate resource?
What if they lose the name/id of this job over the hours to days it takes to complete?

3. As a sub resource of amlfilesystem,
If this becomes a sub resource, since this is an operation, not a setting, it should go away at some point so they don't build up. 
Can these sub resource's be auto deleted at some point after they reach a terminal state of complete, cancelled, failed?  
If not, do we block creation of new ones while a specific amount exists forcing them to delete them?

This would be good in that it would make the status array from amlfilesystem GET go away and the list/get on the sub resources would show the details and provide a sort of history up until the auto deletion point.
It could provide the async behavior so that the import can be tracked from the PUT.
It would allow imports to work with ARM templates and other things that work with resources better than POST operations.1

Can an initial sub resource be provided in the amlfilesystem creation PUT to start the import of the data into the cluster?  The current api supports that so we would have to do that in some way.